compiler/IREmitter/Payload.cc

```diff
@@ -985,8 +987,29 @@ llvm::Value *Payload::varGet(CompilerState &cs, cfg::LocalRef local, llvm::IRBui
     }
     if (irctx.escapedVariableIndices.contains(local)) {
         auto *cfp = getCFPForBlock(cs, builder, irctx, rubyBlockId);
-        auto [index, level] = escapedVariableIndexAndLevel(cs, local, irctx, rubyBlockId);
-        return builder.CreateCall(cs.getFunction("sorbet_readLocal"), {cfp, index, level});
+        auto info = escapedVariableInfo(cs, local, irctx, rubyBlockId);
+        auto *value = builder.CreateCall(cs.getFunction("sorbet_readLocal"), {cfp, info.index, info.level});
+        // If we ever wrote to this local, we cannot guarantee that the type has
+        // been checked prior to the access from the locals array.
+        if (info.use.used == LocalUsedHow::WrittenTo) {
+            return value;
+        }
```

Comment:
Do you have a sense for whether or not we have enough static information to relax this in some cases? Since this PR tracks whether or not a variable is ever written to, the optimization would be disabled for cases where a variable is written to once at the end of a function, but is only read inside of a block.